ckan/lib/helpers.py

```diff
@@ -2751,3 +2758,78 @@ def check_ckan_version(min_version: Optional[str] = None,
     """
     return p.toolkit.check_ckan_version(min_version=min_version,
                                         max_version=max_version)
+
+
+def make_login_url(
+    login_view: str, next_url: Optional[str] = None, next_field: str = "next"
+) -> str:
+    '''
+    Creates a URL for redirecting to a login page. If only `login_view` is
+    provided, this will just return the URL for it. If `next_url` is provided,
+    however, this will append a ``next=URL`` parameter to the query string
+    so that the login view can redirect back to that URL.
+    '''
+    base = login_view
+    if next_url is None:
+        return base
+
+    if url_is_local(next_url):
+        md = {}
+        md[next_field] = urlparse(next_url).path
+        parsed_base = urlparse(base)
+        netloc = parsed_base.netloc
+        parsed_base = parsed_base._replace(netloc=netloc, query=urlencode(md))
+        return urlunparse(parsed_base)
+    return base
+
+
+def _ckan_login_required(  # type: ignore
+    func: Any,
+) -> Callable[..., Union[Response, tuple[str, Optional[int]]]]:
+    """
+    This `CKAN` decorator exists only to decorate
+    the `error_handler`.
```

Comment:
Don't you think that would be too much code for one function?  I thought it's better to make a separate function and use it as a decorator that would be configurable via our `.ini` file, so we can turn it off/on depending on our needs.

Reply:
Maybe I'm missing something but I think it's better to have all related logic in the same place:

```diff
diff --git a/ckan/config/middleware/flask_app.py b/ckan/config/middleware/flask_app.py
index 874bef9c4..f64d41e6f 100644
--- a/ckan/config/middleware/flask_app.py
+++ b/ckan/config/middleware/flask_app.py
@@ -17,7 +17,12 @@ from flask.ctx import _AppCtxGlobals
 from flask.sessions import SessionInterface
 from flask_multistatic import MultiStaticFlask
 
-from werkzeug.exceptions import default_exceptions, HTTPException
+from werkzeug.exceptions import (
+    default_exceptions,
+    HTTPException,
+    Unauthorized,
+    Forbidden
+)
 from werkzeug.routing import Rule
 from werkzeug.local import LocalProxy
 
@@ -33,7 +38,7 @@ from ckan.lib import helpers
 from ckan.lib import jinja_extensions
 from ckan.lib import uploader
 from ckan.lib import i18n
-from ckan.common import config, g, request, ungettext
+from ckan.common import config, g, request, ungettext, current_user
 from ckan.config.middleware.common_middleware import (TrackingMiddleware,
                                                       HostHeaderMiddleware,
                                                       RootPathMiddleware)
@@ -521,6 +526,25 @@ def _register_error_handler(app: CKANApp):
 
     @helpers._ckan_login_required
     def error_handler(e: Exception) -> tuple[str, Optional[int]]:
+
+        if not config.get_value("ckan.login_required.disabled"):
+            # if the url is not local we dont want to redirect the user
+            # instead, we want to show the actual 403(Forbidden)...
+            # same for user.perform_reset if the current_user.is_deleted()
+            # the view returns 403 hence we want to show the actual exception.
+            endpoints = tuple(["util.internal_redirect", "user.perform_reset"])
+            if request.endpoint not in endpoints:
+                if isinstance(e, HTTPException):
+                    if current_user.is_anonymous and type(e) in (Unauthorized, Forbidden):
+                        # make next_url
+                        # TODO: Any reason not to use url_for("user.login") here?
+                        login_url = helpers.redirect_to("user.login").headers["location"]
+                        next_url = request.url
+                        redirect_url = helpers.make_login_url(login_url, next_url=next_url)
+
+                        helpers.flash("Please log in to access this page.")
+                        return helpers.redirect_to(redirect_url)
+
         debug = config.get_value('debug')
         if isinstance(e, HTTPException):
             if debug:

```

Reply:
Yes, this is much better, i've made only one minor change to avoid double `if isinstance(e, HTTPException):` in the code, please see my latest commit and let me know if you agree to this.

Thanks!